src/coreclr/jit/fginline.cpp

```diff
@@ -586,8 +606,59 @@ class SubstitutePlaceholdersAndDevirtualizeWalker : public GenTreeVisitor<Substi
                 }
 
                 CORINFO_CONTEXT_HANDLE contextInput = context;
+                context                             = nullptr;
                 m_compiler->impDevirtualizeCall(call, nullptr, &method, &methodFlags, &contextInput, &context,
                                                 isLateDevirtualization, explicitTailCall);
+
+                if (!call->IsVirtual())
+                {
+                    assert(context != nullptr);
+                    CORINFO_CALL_INFO callInfo = {};
+                    callInfo.hMethod           = method;
+                    callInfo.methodFlags       = methodFlags;
+                    m_compiler->impMarkInlineCandidate(call, context, false, &callInfo);
+
+                    if (call->IsInlineCandidate())
+                    {
+                        Statement* newStmt = nullptr;
+                        GenTree**  callUse = nullptr;
+                        if (m_compiler->gtSplitTree(m_compiler->compCurBB, m_curStmt, call, &newStmt, &callUse, true))
+                        {
+                            if (m_firstNewStmt == nullptr)
+                            {
+                                m_firstNewStmt = newStmt;
+                            }
+                        }
+
+                        // If the call is the root expression in a statement, and it returns void,
+                        // we can inline it directly without creating a RET_EXPR.
+                        if (parent != nullptr || call->gtReturnType != TYP_VOID)
+                        {
+                            Statement* stmt = m_compiler->gtNewStmt(call);
+                            m_compiler->fgInsertStmtBefore(m_compiler->compCurBB, m_curStmt, stmt);
+                            if (m_firstNewStmt == nullptr)
+                            {
+                                m_firstNewStmt = stmt;
+                            }
+
+                            GenTreeRetExpr* retExpr =
+                                m_compiler->gtNewInlineCandidateReturnExpr(call->AsCall(),
+                                                                           genActualType(call->TypeGet()));
+                            call->GetSingleInlineCandidateInfo()->retExpr = retExpr;
+
+                            JITDUMP("Creating new RET_EXPR for [%06u]:\n", call->gtTreeID);
+                            DISPTREE(retExpr);
+
+                            *pTree = retExpr;
+                        }
+
+                        call->GetSingleInlineCandidateInfo()->exactContextHandle = context;
+                        INDEBUG(call->GetSingleInlineCandidateInfo()->inlinersContext = call->gtInlineContext);
```

Comment:
@hez2010 I think this Debug logic is causing Debug and Release JITs to report differing amounts of metadata to SuperPMI. [runtime-coreclr superpmi-asmdiffs-checked-release](https://dev.azure.com/dnceng-public/public/_build?definitionId=153&_a=summary) broke around the time this PR was merged, and I'm not getting any diffs locally when using builds from the commit before this one.

cc @AndyAyersMS @jakobbotsch. I'm not sure what the ideal fix is here, since `GenTreeCall::gtInlineContext` is Debug-only, and removing this logic altogether triggers Debug asserts in `InlineStrategy::NewContext`.

Reply:
I think we may need to move `call->gtInlineContext` out of `#if DEBUG` and always set `call->GetSingleInlineCandidateInfo()->inlinersContext` accordingly.

Reply:
Can the information be put into `LateDevirtualizationInfo`?

Reply:
We have avoided trying to add more fields to GenTreeCall that are only needed for some small subset of all calls.

What @jakobbotsch suggests makes sense to me.

Reply:
Okay I revised this issue and found that the issue is that the debug logic should really be

```cpp
INDEBUG(call->gtInlineContext = call->GetSingleInlineCandidateInfo()->inlinersContext);
```
instead.

We set a new `inlinersContext` for the inline candidate, but didn't update the original `InlineContext` for the call.